### PR TITLE
Add backend field to TrafficMetrics object

### PR DESCRIPTION
**Pull request by adleong**
In order to support doing SMI metrics queries for traffic split resources, we add a `backend` field onto the TrafficMetrics object.  When present, this indicates which branch of the traffic split these metrics apply to.  

For example, this will facilitate responses like this:

```
{
  "kind": "TrafficMetricsList",
  "apiVersion": "metrics.smi-spec.io/v1alpha1",
  "metadata": {
    "selfLink": "/apis/metrics.smi-spec.io/v1alpha1/namespaces/emojivoto/trafficsplits"
  },
  "resource": {
    "kind": "TrafficSplit",
    "namespace": "emojivoto",
  },
  "items": [
    {
      "kind": "TrafficMetrics",
      "apiVersion": "metrics.smi-spec.io/v1alpha1",
      "metadata": {
        "name": "webapp-split",
        "namespace": "emojivoto",
        "selfLink": "/apis/metrics.smi-spec.io/v1alpha1/namespaces/emojivoto/trafficsplits/webapp-split"
        "creationTimestamp": "2020-02-18T23:10:11Z"
      },
      "timestamp": "2020-02-18T23:10:11Z",
      "window": "30s",
      "resource": {
        "kind": "TrafficSplit",
        "namespace": "emojivoto",
        "name": "webapp-split"
      },
      "edge": {
        "direction": "from",
        "resource": null,
      },
        "backend": {
          "apex": "web-svc",
          "name": "emoji-svc",
          "weight": 100,
        }
      "metrics": [
        {
          "name": "p99_response_latency",
          "unit": "ms",
          "value": "4800m"
        },
        {
          "name": "p90_response_latency",
          "unit": "ms",
          "value": "3909m"
        },
        {
          "name": "p50_response_latency",
          "unit": "ms",
          "value": "3181m"
        },
        {
          "name": "success_count",
          "value": "57"
        },
        {
          "name": "failure_count",
          "value": "3"
        }
      ]
    },
    {
      "kind": "TrafficMetrics",
      "apiVersion": "metrics.smi-spec.io/v1alpha1",
      "metadata": {
        "name": "webapp-split",
        "namespace": "emojivoto",
        "selfLink": "/apis/metrics.smi-spec.io/v1alpha1/namespaces/emojivoto/trafficsplits/webapp-split"
        "creationTimestamp": "2020-02-18T23:10:11Z"
      },
      "timestamp": "2020-02-18T23:10:11Z",
      "window": "30s",
      "resource": {
        "kind": "TrafficSplit",
        "namespace": "emojivoto",
        "name": "webapp-split"
      },
      "edge": {
        "direction": "from",
        "resource": null,
      },
        "backend": {
          "apex": "web-svc",
          "name": "web-svc",
          "weight": 900,
        }
      "metrics": [
        {
          "name": "p99_response_latency",
          "unit": "ms",
          "value": "4800m"
        },
        {
          "name": "p90_response_latency",
          "unit": "ms",
          "value": "3909m"
        },
        {
          "name": "p50_response_latency",
          "unit": "ms",
          "value": "3181m"
        },
        {
          "name": "success_count",
          "value": "0"
        },
        {
          "name": "failure_count",
          "value": "12"
        }
```

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

File: pkg/apis/metrics/v1alpha2/edge.go [hunk 1]
```diff
@@ -26,9 +26,10 @@ const (
 
 // Edge describes the other resource that metrics are associated with
 type Edge struct {
-	Direction Direction           `json:"direction"`
-	Side      Side                `json:"side"`
-	Resource  *v1.ObjectReference `json:"resource"`
+	Direction    Direction           `json:"direction"`
+	Side         Side                `json:"side"`
+	Resource     *v1.ObjectReference `json:"resource"`
+	TrafficSplit *TrafficSplit       `json:"traffic_split"`
```
Comment (stefanprodan) [inline pkg/apis/metrics/v1alpha2/edge.go]:
```suggestion
	TrafficSplit *TrafficSplit       `json:"trafficSplit"`
```

File: pkg/apis/metrics/v1alpha2/traffic_split.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package v1alpha2
+
+// TrafficSplit describes which branch of a traffic split these metrics
+// represent.
+type TrafficSplit struct {
+	apex   string `json:"apex"`
+	leaf   string `json:"leaf"`
+	weight int    `json:"weight"`
```
Comment (stefanprodan) [inline pkg/apis/metrics/v1alpha2/traffic_split.go]: Reusing `TrafficSplit` name is very confusing to me, shouldn't this be reference to a existing TrafficSplit? something like:
```
type TrafficSplit struct {
*v1.ObjectReference
Backend string
}
```

What about HTTP match conditions?

Comment (adleong) [reply pkg/apis/metrics/v1alpha2/traffic_split.go]: You're right, this is a confusing name.  This object is meant to indicate which branch of the traffic of traffic split these metrics represent.  Do you have any suggestions for a better name?  What about something like, `traffic_split_branch`, or `backend`?

**Comment by adleong** [comment]
Thanks for taking a look at this, @stefanprodan!  The motivation is to allow this API to return metrics about a traffic split resource.  In particular, we would like to be able to see metrics for each branch of the traffic split.  See the description of this PR for an example of how that API response could look.

I will prepare a corresponding PR for the smi-spec repo.  Unfortunately, that repo only contains a somewhat vague description of the API and no formal specification.  As far as I can tell, the only formal specification for this API exists in this repo which is why I made the change here first.

File: pkg/apis/metrics/v1alpha2/traffic_split.go (hunk 2 above)
Comment (stefanprodan) [reply pkg/apis/metrics/v1alpha2/traffic_split.go]: I think `backend` would be better.

**Comment by adleong** [comment]
https://github.com/deislabs/smi-spec/pull/115

File: pkg/apis/metrics/v1alpha2/backend.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package v1alpha2
+
+// Backend describes which branch of a traffic split these metrics
+// represent.
+type Backend struct {
+	apex   string `json:"apex"`
+	leaf   string `json:"leaf"`
```
Comment (stefanprodan) [inline pkg/apis/metrics/v1alpha2/backend.go]: Instead of backend.leaf could be backend.name?

**Comment by adleong** [comment]
I have move the "backend" struct from the "edge" object to the "trafficmetrics" object.  This feels more natural since the branch of the traffic split doesn't really have anything to do with which edge this is.

**Comment by adleong** [comment]
@stefanprodan thanks! done!

**Pull request merged**